### PR TITLE
[FIX] website: handle device visibility on table of content

### PR DESCRIPTION
**Pull request by Guillaume-gdi**
By following these steps:
- Drop a block table of content on a page
- Disable the visibility of the block in desktop view

=> A traceback is displayed.
This is because for scrollspy to work properly, the elements that it
handles must be visible (no display: none). Unfortunately, we put a
display none when the block must be invisible for a certain device and
the width of the screen is the one of this device. Since it is quite
complex to prevent all the cases where the block could become invisible,
we patch the scrollspy component of bootstrap so that it has a similar
behavior as in version 4.X. (not cause an error if the navigation
element is no longer in the DOM or is no longer visible).

The error is only visible since [the migration from bootstrap 4 to
bootstrap 5] because to add the class, bootstrap 4 did it with the
JQuery addClass() function which does not cause an error if the element
on which it is called does not exist. Now, bootstrap 5 does the same
thing in pure JS with classList.add() on elements that meet the
requirements (visible) which causes an error if the element on which it
is called is not defined (this is the case before this commit because
the TOC was not visible).

---

From [this commit], it is possible to define the visibility of blocks
according to the device on which the page is displayed (either hide on
mobile, or hide on desktop). This change has brought two problems on the
table of content block:
1. The sub-component s_table_of_content_main (the right part) can be
invisible, which does not make sense.
2. The sub-component of s_table_of_content_main can be invisible but it
did not change the navbar of the TOC.

This commit prevents hiding the main part of the TOC and rebuilds the
navbar of the TOC according to the device visibility of its
sub-components.

---

Since [this commit], it is possible to have a block only displayed on
mobile screens. This feature can break the steps snippet by following
these steps:
- Drop a steps snippet in a page
- Set the third step to be visible only on mobile

=> The connectors are broken.
This commit fixes this issue by only considering steps that are visible
in desktop view for connectors drawing.

---

[the migration from bootstrap 4 to bootstrap 5]: https://github.com/odoo/odoo/commit/c48f57ea2538ad51e00ac27d58f8e191781444f3
[this commit]: https://github.com/odoo/odoo/commit/3103e0553011b5c1f4078972d7a88fa3fd4068b2

task-3116227
opw-3135927

### CODE REVIEW

**Comment by Guillaume-gdi** [comment]
@qsm-odoo This OPW fix is ready for your review :heavy_check_mark:

File: addons/website/static/src/js/editor/snippets.options.js [hunk 1]
```diff
@@ -2489,6 +2489,12 @@ options.registry.DeviceVisibility = options.Class.extend({
      * @see this.selectClass for parameters
      */
     async toggleDeviceVisibility(previewMode, widgetValue, params) {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2491]: If my eyes do not deceive me, there is more than 72 characters per line in your commit message :)

File: addons/website/static/src/js/utils.js [hunk 2]
```diff
@@ -404,6 +404,35 @@ function generateGMapLink(dataset) {
         + '&ie=UTF8&iwloc=&output=embed';
 }
 
+/**
+ * @param  {Object} self
+ * @param  {Function} exception
+ */
```
Comment (qsm-odoo) [inline addons/website/static/src/js/utils.js:410]: Double space error :blush:

File: addons/website/static/src/snippets/s_table_of_content/000.js [hunk 3]
```diff
@@ -18,6 +19,22 @@ const TableOfContent = publicWidget.Widget.extend({
...
+                    wUtils.activateScrollSpy(this,
+                        (tocEl) => tocEl === this.$target[0].closest('.s_table_of_content'));
+                    this.wasVisible = false;
+                } else {
+                    this.wasVisible = true;
+                    this._updateTableOfContentNavbarPosition();
+                }
+            });
+        });
+        this.resizeObserver.observe(this.$target[0]);
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_table_of_content/000.js:36]: In this end, I am not sure about the this need of having to think about table of content everywhere some visibility value changes... :grimacing: Same for the need of a resize observer for the table of content, seems overkill. You could probably just listen to the window resize via the utils of config.js / something like that.

But mainly: is there no way to properly patch the lib (the jquery change you explain in your commit message) to avoid all of this? Maybe even check if they did not solve the issue in a further update of bootstrap, in which case we could really patch the current lib in place?

File: addons/website/static/src/snippets/s_table_of_content/options.js [hunk 4]
```diff
@@ -83,7 +84,8 @@ options.registry.TableOfContent = options.Class.extend({
         // We dispose the scrollSpy because the navbar will be updated.
         wUtils.disposeScrollSpy(this);
         const $nav = this.$target.find('.s_table_of_content_navbar');
-        const $headings = this.$target.find(this.targetedElements);
+        const $headings = this.$target.find(this.targetedElements).toArray()
+            .filter(el => !el.closest('.o_snippet_desktop_invisible'));
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_table_of_content/options.js:88]: I did not tested but here you are in options.js, meaning the navbar of table of content is generated during edit mode... are you sure this makes sense and work outside of edit mode as a visitor? = the navbar displays the section which are currently visible on the current device's size? I don't understand how that can work without replicating the d-none/d-md-xxx classes on the navbar links.

File: addons/website/static/src/snippets/s_table_of_content/000.js [hunk 5]
```diff
@@ -68,15 +67,12 @@ const TableOfContent = publicWidget.Widget.extend({
         this.$target.css('top', isHorizontalNavbar ? position : '');
         this.$target.find('.s_table_of_content_navbar').css('top', isHorizontalNavbar ? '' : position + 20);
         position += isHorizontalNavbar ? this.$target.outerHeight() : 0;
-        if (this.previousPosition !== position) {
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_table_of_content/000.js:54]: So here the commit you revert creates a bug in 16.0 but not in previous versions where it was merged?

File: addons/website/static/src/snippets/s_process_steps/options.js [hunk 6]
```diff
@@ -81,7 +81,7 @@ options.registry.StepsConnector = options.Class.extend({
     _reloadConnectors() {
         const possibleTypes = this._requestUserValueWidgets('connector_type')[0].getMethodsParams().optionsPossibleValues.selectClass;
         const type = possibleTypes.find(possibleType => possibleType && this.$target[0].classList.contains(possibleType)) || '';
-        const steps = this.$target[0].querySelectorAll('.s_process_step');
+        const steps = this.$target[0].querySelectorAll('.s_process_step:not(.o_snippet_desktop_invisible)');
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_process_steps/options.js:84]: I really have the same worries here: how can this work outside of edit mode?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/static/src/snippets/s_table_of_content/options.js (hunk 4 above)
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_table_of_content/options.js:79]: The nav bar of the table of content is not displayed in mobile view so we can ignore the headings visible only in mobile view when we build the nav. I add a comment to explain that :+1:

File: addons/website/static/src/snippets/s_table_of_content/000.js (hunk 5 above)
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_table_of_content/000.js]: No because the mobile preview was not available before 16.0. But as we will finally patch the lib, I will not have to disable the scrollspy and so we can keep the commit :+1:

File: addons/website/static/src/snippets/s_process_steps/options.js (hunk 6 above)
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_process_steps/options.js:84]: It's quite the same as the other one. As the connectors are only for the desktop view we can ignore steps that are only display in mobile view :+1: I will add a comment :)

File: addons/website/static/src/snippets/s_table_of_content/000.js (hunk 3 above)
Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_table_of_content/000.js]: Indeed it is way more simple and robust to go with a lib patch :+1:

**Comment by Guillaume-gdi** [comment]
@qsm-odoo The PR is ready for your review :heavy_check_mark:

**Comment by Guillaume-gdi** [comment]
I just removed the useless `run: 'click'`

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@Guillaume-gdi I did this to test your first commit:
- Take a current 16.0, run the `test_08_table_of_content` test -> it's working :+1: 
- Add the test update made by your first commit, run the test again -> it's crashing, good :+1: 
- Add the fix you made in first commit (lib patch), run the test again -> it's still crashing :cry: 

... maybe you are making the test pass via the two extra commits but this is not ok: each commit should work on its own. So if that is the case, adapt the test in all 3 commits with what you are actually fixing, not just the first one. Can you double check? Thanks :+1:

**Comment by Guillaume-gdi** [comment]
@qsm-odoo Mhhh there is something strange... I did test it before and the test worked with the fix...
I just tested it again 10 times on my laptop, with only the first commit (the fix + the tests) -> It passed
BVR tested it on his laptot -> It passed
I even did a runbot test with the first commit only and it passed ([here](https://github.com/odoo/odoo/pull/111118))...

If you are 100% sure you had an error, maybe there is a race condition that I don't see :/ Can we ask for a multi test of [the branch](https://runbot.odoo.com/runbot/bundle/16-0-test-toc-test-runbot-gdi-145494) with only on the first commit ?

**Comment by qsm-odoo** [comment]
@Guillaume-gdi Strange I retested and here is what I got:

1. Checkout a rebased 16.0 version
2. Cherry-pick your first commit on it
3. Create a database, -i website in console
4. Close the odoo server and run it again with the command `--test-tags=.test_08_table_of_content`
5. Crash as I got before:

```
2023-01-26 17:15:16,130 90974 ERROR odoo_community odoo.addons.website.tests.test_snippets: FAIL: TestSnippets.test_08_table_of_content
Traceback (most recent call last):
  File "/home/qsm/Documents/code/src/odoo/addons/website/tests/test_snippets.py", line 54, in test_08_table_of_content
    self.start_tour(self.env['website'].get_client_action_url('/'), 'snippet_table_of_content', login='admin')
  File "/home/qsm/Documents/code/src/odoo/odoo/tests/common.py", line 1999, in start_tour
    return self.browser_js(url_path=url_path, code=code, ready=ready, **kwargs)
  File "/home/qsm/Documents/code/src/odoo/odoo/tests/common.py", line 1979, in browser_js
    self.fail('%s\n\n%s' % (message, error))
AssertionError: The test code "odoo.startTour('snippet_table_of_content')" failed

Tour snippet_table_of_content failed at step Hide the first TOC on mobile (trigger: [data-toggle-device-visibility="no_mobile"])
```

With the associated screenshot:
image

As you can see the step
```javascript
    {
        content: "Click on the first TOC's title",
        trigger: 'iframe .s_table_of_content:eq(0) h1',
    },
```
Does not seem to have an effect: the table of content options did not appear.

6. I then restart the test to see if it is systematic: no error. I relaunched it 4-5 times -> no error.
7. Dropdb, restart from step 3 => the same occurs: fails the first time the test is launched, works the times afterwards.

I guess it could be ignored (and reported to the runbot team) if runbot does not see the bug and you all don't see that either but it is worth looking into a bit more. Can you try that exact flow :point_up: and tell me if you got the same on your side ? Maybe also @rdeodoo ? :)

**Comment by Guillaume-gdi** [comment]
@qsm-odoo I did multiple times your flow and test succeed every times... I wonder why it works for me and not for you :thinking:

File: addons/website/static/tests/tours/snippet_table_of_content.js [hunk 7]
```diff
@@ -38,4 +38,46 @@ wTourUtils.registerWebsitePreviewTour('snippet_table_of_content', {
...
+        trigger: 'iframe .s_table_of_content.o_snippet_mobile_invisible',
+        run: () => {}, // This is a check.
+    },
+    {
+        content: "Check that the mobile TOC is not visible on desktop",
+        trigger: 'iframe .s_table_of_content:not(.o_snippet_desktop_invisible)',
+        run: () => {}, // This is a check.
+    },
+    {
+        content: "Toggle de mobile view",
```
Comment (qsm-odoo) [inline addons/website/static/tests/tours/snippet_table_of_content.js:70]: Spreek je nu Nederlands? :grimacing:

File: addons/website/static/tests/tours/snippet_table_of_content.js [hunk 8]
```diff
@@ -38,4 +38,46 @@ wTourUtils.registerWebsitePreviewTour('snippet_table_of_content', {
...
+    {
+        content: "Check that we have the good TOC on mobile",
+        trigger: 'iframe .s_table_of_content.o_snippet_desktop_invisible',
+        run: () => {}, // This is a check.
+    },
+    {
+        content: "Check that the desktop TOC is not visible on mobile",
+        trigger: 'iframe .s_table_of_content:not(.o_snippet_mobile_invisible)',
+        run: () => {}, // This is a check.
+    },
```
Comment (qsm-odoo) [inline addons/website/static/tests/tours/snippet_table_of_content.js:82]: I don't understand those steps. The o_snippet_XXX_invisible classes are not the ones who determine if one's is visible or not, but ok to check them in tests. However, what you do here does not seem ok for multiple reasons:
```javascript
    {
        content: "Check that the mobile TOC is not visible on desktop",
        trigger: 'iframe .s_table_of_content:not(.o_snippet_desktop_invisible)',
        run: () => {}, // This is a check.
    },
```
That trigger is always true no matter what. Of course there is a table_of_content that does not have this class if you do not target a specific table of content :thinking:
Correct me if I am wrong but those classes do not change whether or not we are in mobile mode/preview or not... so you basically are checking the same thing outside and in preview mode :thinking: In the end, either you toggle the mobile preview and you check the real visibility of the items on and off mobile view, or you solely rely on the o_snippet_XXX_invisible classes... but mixing both actually does not test this properly.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 9]
```diff
@@ -2509,6 +2509,7 @@ options.registry.DeviceVisibility = options.Class.extend({
             },
         });
         this.trigger_up('snippet_option_visibility_update', {show: widgetValue !== (isMobile ? 'no_mobile' : 'no_desktop')});
+        this.$target.trigger('content_changed');
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:2512]: Why is that necessary? It is normally automatic at the end of all snippet option method calls?
```javascript
        // We trigger the event on elements targeted by apply-to if any as
        // this.$target could not be in an editable element while the elements
        // targeted by apply-to are.
        ($applyTo || this.$target).trigger('content_changed');
```

File: addons/website/static/src/snippets/s_table_of_content/options.js [hunk 10]
```diff
@@ -72,7 +73,10 @@ options.registry.TableOfContent = options.Class.extend({
         // We dispose the scrollSpy because the navbar will be updated.
         this._disposeScrollSpy();
         const $nav = this.$target.find('.s_table_of_content_navbar');
-        const $headings = this.$target.find(this.targetedElements);
+        // As the navbar is hidden in mobile view, we only consider the headings
+        // visible on desktop.
+        const $headings = this.$target.find(this.targetedElements).toArray()
+            .filter(el => !el.closest('.o_snippet_desktop_invisible'));
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_table_of_content/options.js:79]: If I am not mistaken, after this, `$headings` is not a jQuery object anymore and should thus not be named that way anymore.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
I tried 2 times again about the test that does not work on my side, following the flow I described earlier... and did not get the error anymore :shrug: Strange. Let's forget about it for now. Some new requested changes :wink:

File: addons/website/static/tests/tours/snippet_table_of_content.js (hunk 7 above)
Comment (Guillaume-gdi) [reply addons/website/static/tests/tours/snippet_table_of_content.js]: Geen woord... waarschijnlijk een fonetische schrijfreflex... Oeps :grimacing:

**Comment by Guillaume-gdi** [comment]
@qsm-odoo I adapted to your comments :heavy_check_mark: 
Let me know if there is still something I can do about this one. I can still adapt tomorrow if needed :+1:

**Comment by Guillaume-gdi** [comment]
Just resolve the conflict :+1: 

@qsm-odoo I ping you so that we don't forget this PR.

**Review by qsm-odoo — APPROVED** [review]
@robodoo rebase-ff r+ :+1:
